drivers: video: sw_generator: add more formats and try to simplify

josuah opened the pull request:
This includes several incremental, small changes to the software generator.

- drivers: video: sw_generator: modify video_sw_generator_fill_colorbar()
- drivers: video: sw_generator: refactor fmt/frmival selection
- drivers: video: sw_generator: refactor: flatten arrays
- drivers: video: sw_generator: preserve full prefix for internal functions

It was not broken on ideal conditions: this only help it be more robust during misuse.

josuah committed:
drivers: video: sw_generator: preserve full prefix for internal functions

In order to help debugging through GDB and other error messages and debug
tools, convert the __xxx prefix to video_sw_generator_xxx full prefix.
To help keep function names short, use slightly shorter sufixes.

Signed-off-by: Josuah Demangeon <me@josuah.net>

josuah committed:
drivers: video: sw_generator: refactor: flatten arrays

Help with maintainance and possibly readability by using a more regular
layout for various tables of numbers. This adds a comma on the last
element to help with formatters like clang-format.

Signed-off-by: Josuah Demangeon <me@josuah.net>

josuah committed:
drivers: video: sw_generator: fix video_sw_generator_enum_frmival()

Return an error on fie.index >= 1 as there is only one framerate entry
per pixelformat, this prevents an endless loop.

Signed-off-by: Josuah Demangeon <me@josuah.net>

josuah committed:
drivers: video: sw_generator: modify video_sw_generator_fill_colorbar()

Add a check for the array size to avoid overwriting unrelated memory when
the buffer is too small for the full format. It first check if there is
enough buffer for one line, and fill it programmatically. Then, it will
try to duplicate that line over the entire buffer, in the limit of the
room available.

Signed-off-by: Josuah Demangeon <me@josuah.net>

josuah committed:
drivers: video: sw_generator: add support for bayer and YUV formats

This refactors the pattern generator functions to also offer bayer formats
as input. 4 variant of bayer formats are proposed. The pixel packing is
also now split from the color selection: only a single RGB and single YUV
array used by all the pattern generators.

Signed-off-by: Josuah Demangeon <me@josuah.net>

josuah commented:
force-push:
- Added 4 different bayer formats
- Added YUYV format
- Refactored the pattern generator loop

This is a draft until I can properly test every format, i.e. with UVC or other ways.

Ideally this would be having unit tests like done by the capture sample, but this would need to come in another PR.

josuah committed:
drivers: video: sw_generator: introduce the RGB24 test pattern

This complements the 32-bit RGB (XRGB32) test pattern with an equivalent
24-bit RGB (RGB24) implementation.

Signed-off-by: Josuah Demangeon <me@josuah.net>

josuah committed:
drivers: video: sw_generator: make k_work_sync a local variable

The documentation of k_work_cancel_delayable_sync() states that the input
k_work_sync parameter needs to be valid until the function call returns,
so there is no need to preserve the state across successive calls.
Now that there is a single work-related field in the struct, rename it
to simply "work".

Signed-off-by: Josuah Demangeon <me@josuah.net>

josuah commented:
Force-push:
- The following formats could be tested with external tooling, not part of upstream Zephyr: YUYV, RGB24, RGGB8
- Simplification of how `struct k_work_sync` is used

There is a challenge though: the YUYV is actually YUV BT.709 (used by sRGB and HDTV), not BT.601 (used by JPEG).
So using the pattern generator and piping it into a JPEG library gives a test pattern with wrong contrast.

josuah commented:
force-push:
- Test the formats using the lib/pixel, and bugfix them.
- A bit more input validation

ngphibang commented on drivers/video/video_sw_generator.c:
```diff
@@ -131,25 +131,40 @@ static void video_sw_generator_fill_colorbar(struct video_sw_generator_data *dat
 		}
+
+		memcpy(vbuf->buffer + h * data->fmt.pitch, vbuf->buffer, data->fmt.pitch);
+		vbuf->bytesused += data->fmt.pitch;
```
I think the old code is OK. I think we don't need to check the buffer against the format's pitch because

- In the current v4z framework, when application allocates the video buffer, it is supposed to allocate sufficient memory for the requested format, if not, it will simply not work.
- Otherwise, there is a problem in the current framework which will be pointed out in my upcomming buffer management PR: 

1. the application should not specify the format->pitch as it has no knowdlege about this (this should be set by the driver, e.g. driver can add padding to the width line, etc.)
2. the application also should not specify the buffer size, it just set the format, the number of buffers and request the framework / driver to alloc the buffers on the video buffer heap.

josuah replied:
I replaced `data->fmt.pitch` by a local `pitch` variable that is computed locally: the driver knows what it generates, let me know if that does not solve the concerns fully.

(2.) takes the entire buffer management out of the application developer way! :hushed:

ngphibang commented on drivers/video/video_sw_generator.c:
```diff
@@ -85,6 +85,11 @@ static int video_sw_generator_set_fmt(const struct device *dev, enum video_endpo
 		return -ENOTSUP;
 	}
 
+	if (fmt->pitch != fmt->width * video_bits_per_pixel(fmt->pixelformat) / BITS_PER_BYTE) {
+		LOG_ERR("The pitch is not consistent with the other parameters");
+		return -EINVAL;
+	}
```
Same comment here. This reveals  the weak point of the current framework : application should not set the format's pitch.

josuah replied:
Removed.

ngphibang commented on drivers/video/video_sw_generator.c:
```diff
@@ -38,24 +38,27 @@ struct video_sw_generator_data {
+		.width_step = 1,
+		.height_step = 1,
+	},
+	{0},
```
{0} is the terminator, we still need the comma ? I am not sure if clang-format require a comma on the last element, it seems still pass the clang-format check even without the last comma ? ... just a minor question

josuah replied:
The comma was to help with clang-format, yes. Otherwise, it does something like `{0}};` instead of `{0}\n};`.

ngphibang commented on drivers/video/video_sw_generator.c:
```diff
@@ -31,8 +31,7 @@ struct video_sw_generator_data {
 	struct video_format fmt;
 	struct k_fifo fifo_in;
 	struct k_fifo fifo_out;
-	struct k_work_delayable buf_work;
-	struct k_work_sync work_sync;
+	struct k_work_delayable work;
```
A part from removing work_sync, the commit also rename the buf_work variable which is not claimed in the commit message

josuah replied:
Good catch! I added explanation, I am also 100% fine with preserving the old name.

ngphibang commented on drivers/video/video_sw_generator.c:
```diff
@@ -313,18 +315,20 @@ static int video_sw_generator_enum_frmival(const struct device *dev, enum video_
 {
 	int i = 0;
 
-	if (ep != VIDEO_EP_OUT || fie->index) {
+	if (ep != VIDEO_EP_OUT || fie->index > 0) {
 		return -EINVAL;
 	}
 
-	while (fmts[i].pixelformat && (fmts[i].pixelformat != fie->format->pixelformat)) {
-		i++;
+	for (i = 0; fmts[i].pixelformat != 0; ++i) {
+		if (fie->format->pixelformat == fmts[i].pixelformat &&
+		    IN_RANGE(fie->format->width, fmts[i].width_min, fmts[i].width_max) &&
+		    IN_RANGE(fie->format->height, fmts[i].height_min, fmts[i].height_max)) {
+			break;
+		}
 	}
 
-	if ((i == ARRAY_SIZE(fmts)) || (fie->format->width > fmts[i].width_max) ||
-	    (fie->format->width < fmts[i].width_min) ||
-	    (fie->format->height > fmts[i].height_max) ||
-	    (fie->format->height < fmts[i].height_min)) {
+	if (fmts[i].pixelformat == 0) {
+		LOG_ERR("Nothing matching the requested format was found");
```
I think the change is the same as the old code except the use of IN_RANGE(). But well, both are fine to me.

josuah replied:
This is just an attempt at using idioms.
I reverted `fmts[i].pixelformat == 0` back to `i == ARRAY_SIZE(fmts)` to avoid too many distracting changes.

josuah replied:
I think I forgot to change it there... I notice that maybe `ARRAY_SIZE(fmts)` needs to be `ARRAY_SIZE(fmts) - 1` to exclude the `{0}` terminator from the comparisons.

josuah replied:
For this reason along with [avoiding memory allocation](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/video/video_mcux_csi.c#L355), it could be convenient to switch to `video_enum_format()` like `video_enum_frmival()`. For another RFC/PR maybe!

josuah commented on drivers/video/video_sw_generator.c:
```diff
@@ -10,6 +10,8 @@
 #include <zephyr/drivers/video-controls.h>
 #include <zephyr/logging/log.h>
+#include <zephyr/sys/util.h>
+#include <zephyr/sys/byteorder.h>
```
Note: this change is now in the commit that uses `sys_cpu_to_le32()`. Github moved the comment to a different commit it seems.

ngphibang replied:
Ok, just sort in alphabetical order (nit)

josuah commented on drivers/video/video_sw_generator.c:
```diff
@@ -102,56 +107,159 @@ static int video_sw_generator_get_fmt(const struct device *dev, enum video_endpo
+};
+
+static int video_sw_generator_fill_yuyv(uint8_t *buffer, uint16_t width)
 {
```
Let's increase assert density then!

josuah committed:
drivers: video: sw_generator: use video_format_caps_index()

Use the video_format_caps_index() function to check if a format matches
any entry of the format caps.

Signed-off-by: Josuah Demangeon <me@josuah.net>

josuah commented:
force-push:
- Split the commit [discussed here](https://github.com/zephyrproject-rtos/zephyr/pull/85968#discussion_r1999418703) in two
- Add extra assert() as suggested by bot review [here](https://github.com/zephyrproject-rtos/zephyr/pull/85968#discussion_r2050744278)
- rebase on main()

josuah commented:
Force-push:

- Updated on latest main
- Integrated the `hflip` control everywhere

On `native_sim` with the SDL display:

![scrot_20250427_151658_1640x585](https://github.com/user-attachments/assets/da01cdec-8fbe-490f-a895-122aed667d5d)

![scrot_20250427_151725_1559x461](https://github.com/user-attachments/assets/7fb31678-c64c-4f4e-88aa-162892f280ce)

josuah committed:
drivers: video: sw_generator: fix div by zero on very low framerate

When the FPS value stored in data->frame_rate is zero, a division by zero
occurs. Fix it by clamping the frame rate between min and max values.

Signed-off-by: Josuah Demangeon <me@josuah.net>

ngphibang reviewed (commented):
LGTM, just a nit left

josuah commented:
Force-push:

- Rebased on latest main
- No more endpoint IDs
- RGGB8 -> SRGGB8
- Bugfix for the video capture sample
- Minor logging bugfixes.

Built without the Zephyr Shell (coming in another PR):
![scrot_20250518_172916_1366x442](https://github.com/user-attachments/assets/a68488b1-ee58-4371-b0f9-869284f59360)

ngphibang commented on samples/drivers/video/capture/src/main.c:
```diff
@@ -254,14 +254,15 @@ int main(void)
 			return 0;
 		}
 
+		vbuf->type = type;
```
I think the old code is correct. You need to set the type before calling `video_dequeue()` for the driver to know which queue it needs to get the buffer.

josuah replied:
Thank you I was having this doubt about what to do while implementing the shell.

I will modify to a different fix as `vbuf` is uninitialized on first call:

https://github.com/zephyrproject-rtos/zephyr/blob/b5934e8dfe5c8dd4c1a9f2221341aa46e3dce72a/samples/drivers/video/capture/src/main.c#L93

Not a big issue!

ngphibang commented:
In the attached photo above, it seems the default frame rate is not printed (`%f`) ?

josuah committed:
drivers: video: sw_generator: sort header alphabetically

Sort alphabetically the header added in the previous few commits
as well as the original implementation.

Signed-off-by: Josuah Demangeon <me@josuah.net>

josuah committed:
samples: drivers: video: set an initial video buffer

Provide an initial video buffer in the first call to video_dequeue().
It is used by the driver for choosing between the input or output queue.
Applied to capture, capture_to_lvgl and tcpserversink samples.

Signed-off-by: Josuah Demangeon <me@josuah.net>

josuah committed:
samples: drivers: video: capture: logging improvements

Remove trailing spaces at the end of log string.
Move arguments on the next line for readability.
Align arguments to first line.

Signed-off-by: Josuah Demangeon <me@josuah.net>

josuah commented:
Trying on an FRDM-MCXN, I see the correct format string:

```
*** Booting Zephyr OS build v4.1.0-4199-g4011ff68fdc7 ***
[00:00:00.000,000] <inf> main: Video device: VIDEO_SW_GENERATOR
[00:00:00.000,000] <inf> main: - Capabilities:
[00:00:00.000,000] <inf> main:   RGB3 width [64; 1920; 1] height [64; 1080; 1]
[00:00:00.000,000] <inf> main:   YUYV width [64; 1920; 1] height [64; 1080; 1]
[00:00:00.000,000] <inf> main:   RGBP width [64; 1920; 1] height [64; 1080; 1]
[00:00:00.000,000] <inf> main:   BX24 width [64; 1920; 1] height [64; 1080; 1]
[00:00:00.000,000] <inf> main:   RGGB width [64; 1920; 1] height [64; 1080; 1]
[00:00:00.000,000] <inf> main:   GRBG width [64; 1920; 1] height [64; 1080; 1]
[00:00:00.000,000] <inf> main:   BA81 width [64; 1920; 1] height [64; 1080; 1]
[00:00:00.000,000] <inf> main:   GBRG width [64; 1920; 1] height [64; 1080; 1]
[00:00:00.000,000] <inf> main: - Video format: RGBP 320x160
[00:00:00.000,000] <inf> main: - Default frame rate : 30.000000 fps
[00:00:00.000,000] <inf> main: - Supported frame intervals for the default format:
[00:00:00.000,000] <inf> main:    [min = 1/60; max = 4294967295/1; step = 1/1000]
[00:00:00.000,000] <inf> main: - Supported controls:
[00:00:00.000,000] <inf> video_ctrls:                  Horizontal Flip 0x00980914 (bool)   (flags=0x00) : min=0 max=1 step=1 default=0 value=0
uart:~$
```

I was testing on `native_sim/native/64` which has the host libc enabled:

```
# CONFIG_MINIMAL_LIBC is not set
# CONFIG_PICOLIBC is not set
CONFIG_EXTERNAL_LIBC=y
# CONFIG_COMMON_LIBC_ASCTIME_R is not set
# CONFIG_COMMON_LIBC_CTIME_R is not set
# CONFIG_COMMON_LIBC_MALLOC is not set
# CONFIG_COMMON_LIBC_REMOVE is not set
```

There might be some missing snippet somewhere to enable the float library on the host (`native_sim`). Maybe some compiler flag.

Do you wish it to be fixed in that PR? I can open an issue too if that matters.

josuah commented:
Testing on `main` with `native_sim/native/64` I also have this bug:

```
-- west flash: using runner native
uart connected to pseudotty: /dev/pts/0
*** Booting Zephyr OS build v4.1.0-4186-gb5934e8dfe5c ***
[00:00:00.000,000] <inf> main: Video device: VIDEO_SW_GENERATOR
[00:00:00.000,000] <inf> main: - Capabilities:
[00:00:00.000,000] <inf> main:   RGBP width [64; 1920; 1] height [64; 1080; 1]
[00:00:00.000,000] <inf> main:   BX24 width [64; 1920; 1] height [64; 1080; 1]
[00:00:00.000,000] <inf> main: - Video format: RGBP 320x160
[00:00:00.000,000] <inf> main: - Default frame rate : %f fps                    <<< HERE
[00:00:00.000,000] <inf> main: - Supported frame intervals for the default format:
[00:00:00.000,000] <inf> main:    [min = 1/60; max = 4294967295/1; step = 1/1000]
[00:00:00.000,000] <inf> main:    [min = 1/60; max = 4294967295/1; step = 1/1000]
[00:00:00.000,000] <inf> main:    [min = 1/60; max = 4294967295/1; step = 1/1000]
[00:00:00.000,000] <inf> main:    [min = 1/60; max = 4294967295/1; step = 1/1000]
[00:00:00.000,000] <inf> main:    [min = 1/60; max = 4294967295/1; step = 1/1000]
[00:00:00.000,000] <inf> main:    [min = 1/60; max = 4294967295/1; step = 1/1000]
(endless loop fixed by this PR)
```

avolmat-st commented on drivers/video/video_sw_generator.c:
```diff
@@ -123,29 +124,50 @@ uint32_t xrgb32_colorbar_value[] = {
+	for (int w = 0, i = 0; w < data->fmt.width; w++) {
+		int color_idx = video_sw_generator_get_color_idx(w, width, hflip);
+
+		if (data->fmt.pixelformat == VIDEO_PIX_FMT_RGB565) {
```
With the currently support pixelformat this won't change much however I am thinking that here, instead of checking on the pixelformat, check again the bits_per_pixel could be done in order to understand if 8 / 16  / 32 etc access should be done ?
Then the switch case regarding the pixelformat could be done outside of this for look to just get a pointer to the table of ref data.
Again, not important now, would only allow maybe easier addition of further formats later on.

avolmat-st replied:
Oups, already something done in further commits, please forget about this ;)

josuah replied:
I went with a very step-by-step approach ^_^'

ngphibang reviewed (changes requested):
This seems not to be done "vbuf needs to be initialized and vbuf->type needs to be set before `dequeue()`"

avolmat-st commented on drivers/video/video_sw_generator.c:
```diff
@@ -133,38 +140,119 @@ static inline int video_sw_generator_get_color_idx(uint16_t w, uint16_t width, b
-	bool hflip = data->ctrls.hflip.val;
+	__ASSERT(width % 2 == 0, "YUYV pixels always go by pairs");
+
+	for (size_t w = 0; w + 2 <= width; w += 2) {
```
maybe  w < width here seems easier to understand ?

josuah replied:
Now I realize it is overly defensive to set `w + 2 <= width`:

If we end-up with `w = width - 1` (misaligned) at the beginning of the loop for some reason and consume two pixels inside the loop: `width - 1` and `width - 0`, and `width - 0` is one pixel after the end of the buffer.

Not expected to happen in this particular instance, ASSERT()-tested anyway, let's use `w < width`.
Thanks for pointing that out!

avolmat-st commented on drivers/video/video_sw_generator.c:
```diff
@@ -379,13 +380,12 @@ static int video_sw_generator_set_frmival(const struct device *dev, struct video
-		data->frame_rate = MIN(DIV_ROUND_CLOSEST(frmival->denominator, frmival->numerator),
-				       MAX_FRAME_RATE);
-	} else {
+	if (frmival->denominator == 0 && frmival->numerator == 0) {
```
Isn't this only  numerator that can't be 0 ? I mean 1/0 is a valid frame interval right ?

ngphibang commented on samples/drivers/video/capture/src/main.c:
```diff
@@ -90,7 +90,8 @@ static inline void video_display_frame(const struct device *const display_dev,
 int main(void)
 {
-	struct video_buffer *buffers[CONFIG_VIDEO_BUFFER_POOL_NUM_MAX], *vbuf;
+	struct video_buffer vbuf0 = {0};
+	struct video_buffer *buffers[CONFIG_VIDEO_BUFFER_POOL_NUM_MAX], *vbuf = &vbuf0;
```
We can just use one variable and past the address of vbuf to different functions:

`struct video_buffer vbuf = {0};`

josuah replied:
`vbuf` needs to be a pointer but that can be helped by some tweaks: `*vbuf = &(struct video_buffer){};`

ngphibang replied:
Right, I didn't check that it needs double pointer. Thanks.

ngphibang reviewed (commented):
- I think the commit message / title should be changed too as it is different now:

```
samples: drivers: video: set video buffer types later
Set the video buffer types inside the main loop after the buffers are
dequeued from the video device rather than at the top of the loop.
Applied to both capture and capture_to_lvgl samples.
```

- And we should apply the same commit for tcpserversink sample

josuah commented:
Force-push:
- Modified a few commits to make use of utilities from https://github.com/zephyrproject-rtos/zephyr/pull/79482
- Apply the `vbuf = ...` fix to all 3 video driver samples
- Edit commit messages to match
- `frmival->denominator == 0 && frmival->numerator == 0` fix (`&&` -> `||`)

avolmat-st commented on drivers/video/video_sw_generator.c:
```diff
@@ -126,42 +133,140 @@ static inline int video_sw_generator_get_color_idx(uint16_t w, uint16_t width, b
-					     struct video_buffer *vbuf)
+static uint16_t video_sw_generator_fill_yuyv(uint8_t *buffer, uint16_t width, bool hflip)
+{
+	if (width % 2 == 0) {
```
Shouldn't be != 0 here ?

josuah replied:
Yes I noticed that too late... Corrected.

avolmat-st commented on drivers/video/video_sw_generator.c:
```diff
@@ -126,42 +133,140 @@ static inline int video_sw_generator_get_color_idx(uint16_t w, uint16_t width, b
+						      pattern_grbg8_idx);
+		break;
+	default:
+		LOG_WRN("Unsupported pixel format %x, filling with 0x55", data->fmt.pixelformat);
```
Seems to me that this code is actually not reachable since set_format will have blocked any format which is not part of the sw generator caps. Right ?

avolmat-st commented on drivers/video/video_sw_generator.c:
```diff
@@ -126,42 +133,140 @@ static inline int video_sw_generator_get_color_idx(uint16_t w, uint16_t width, b
-	/* Duplicate the first line all over the buffer */
-	for (int h = 1; h < data->fmt.height; h++) {
-		if (vbuf->size < vbuf->bytesused + pitch) {
+	/* How much was filled insofar */
```
insofar -> in so far

avolmat-st commented on drivers/video/video_sw_generator.c:
```diff
@@ -126,42 +133,138 @@ static inline int video_sw_generator_get_color_idx(uint16_t w, uint16_t width, b
+	return 1;
+}
+
+static uint16_t video_sw_generator_fill_bayer(uint8_t *buffer, uint16_t width, bool hflip,
```
maybe this should be named fill_bayer8 or something like that since it produce 8bit unpacked bayer.

josuah replied:
I removed the 8 from `pattern_rgbr8_idx` as these are bit-depth agnostic, and added it to `fill_bayer`.
Thank you!

avolmat-st replied:
Hum, I am not sure about that (bit-depth agnostic) but maybe I misunderstood what is being generated.
For a 8bit Bayer, indeed, that would generate B G G R for example, each letter being here 8bit. But for other bayers (10 - 12 - 14- 16 etc), each letter should be on 2 bytes, as described within the BGGR10 etc format description in video.h.
That is the reason why I mentioned that this is creating a 8bit bayer pattern, not 10-12-14-16 etc. Right ?

avolmat-st replied:
Yeah I think that if addition is done, this is to be done within another PR. Here my point was only to mentions that it was handling 8 bit bayer hence need to clarify that.
I think usually that is unpacked formats (aka 10-12-14 bayers are all stored into 16bits with 0 padding). I don't know if someone has a need to get the still packed version.

avolmat-st reviewed (changes requested):
Very minor points, appart from that LGTM

josuah commented on drivers/video/video_sw_generator.c:
```diff
@@ -274,6 +274,10 @@ static int video_sw_generator_enum_frmival(const struct device *dev, struct vide
 	int idx;
 	int ret;
 
+	if (fie->index >= 1) {
```
Good catch!

josuah commented:
Force-push:

- adjust the numbers: `_fill_bayer()` -> `_fill_bayer8()`, `pattern_rggb8_idx` -> `pattern_rggb_idx`, `rgb/yuv` -> `rgb24/yuv24`
- Only tested with RGB565 for the `capture` and `capture_to_lgl` samples. I wanted to test everything with UVC (which supports bayer8 formats too) but need to rebase it first...

ngphibang commented:
This PR contains the commit https://github.com/zephyrproject-rtos/zephyr/pull/85968/commits/55eb71f687fcf606077bff263e2f6b853024c79d which could be considered as a hot fix. I tried to assigned @avolmat-st with hope that it can be merged quickly. Correct me if I did it wrong.

josuah commented:
@ngphibang apparently, just 3h left to show-up as "ready" in the merge-list:

- https://merge-list.zephyrproject.io/ (at the bottom, Ctrl+F `sw_generator`)
